<h1 align=center style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
    بسکتبالیست‌ها
</font>
</h1>

<p dir=rtl style="direction: rtl;text-align: right;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
        در این سوال قصد داریم تا داده‌های مربوط به عملکرد بازیکنان لیگ بسکتبال NBA را تحلیل کنیم.
        پس کافی است تا کار‌هایی که از شما خواسته شده را مرحله به مرحله انجام دهید.
    </font>
</p>

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
    <font face="vazir" color="#0099cc">
        مجموعه‌داده
    </font>
</h2>

<p dir=rtl style="direction: rtl;text-align: right;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
        مجموعه‌داده‌ای که در احتیار شما قرار گرفته است شامل اطلاعات عملکرد بازیکنان NBA در فصل‌های ۲۰۰۰ الی ۲۰۰۹ است.
    این مجموعه شامل ۱۸۳۰ سطر و ۱۸ ستون می‌باشد که هر سطر آن شامل اطلاعات یک بازیکن در یک فصل است. توضیحات مربوط به ستون‌ها نیز در جدول زیر آمده است.
    </font>
</p>

<center>
<div dir=rtl style="direction: rtl;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|ستون|توضیحات|
|:------:|:---:|
|<code>PlAYER</code>|نام بازیکن|
|<code>TEAM</code>|مخفف اسم تیم بازیکن در آن فصل|
|<code>YEAR</code>|سال (فصل مسابقات)|
|<code>GP</code>|تعداد بازی‌های انجام شده|
|<code>MIN</code>|میانگین دقایق بازی کرده|
|<code>PTS</code>|میانگین امتیاز|
|<code>FGM</code>|میانگین تعداد پرتاب‌های ۲ امتیازی موفق|
|<code>FGA</code>|میانگین تعداد کل پرتاب‌های ۲ امتیازی|
|<code>3PM</code>|میانگین تعداد پرتاب‌های ۳ امتیازی موفق|
|<code>3PA</code>|میانگین تعداد کل پرتاب‌های ۳ امتیازی|
|<code>FTM</code>|میانگین تعداد پرتاب‌های آزاد (پنالتی) موفق|
|<code>FTA</code>|میانگین تعداد کل پرتاب‌های آزاد (پنالتی)|
|<code>OREB</code>|میانگین تعداد ریباند‌ها در حمله|
|<code>DREB</code>|میانگین تعداد ریباند‌ها در دفاع|
|<code>AST</code>|میانگین تعداد پاس‌های منجر به امتیاز|
|<code>STL</code>|میانگین تعداد دفعات توپ‌ربایی|
|<code>BLK</code>|میانگین تعداد بلاک‌ها|
|<code>TOV</code>|میانگین تعداد دفعات لو دادن توپ|
    
</font>
</div>
</center>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazir; font-size:medium">
<font face="vazir" size=3>
   با استفاده از سلول زیر کتابخانه‌های مورد نیاز خود را <code>import</code> کنید و سپس مجموعه‌داده را که در فایلی به نام <code>NBA2000-2009.csv</code> قرار دارد، بخوانید. 
</font>
</p>

In [28]:
import pandas as pd

In [29]:
df = pd.read_csv('../Data/NBA2000-2009.csv')
df.head()

PLAYER TEAM  YEAR  GP   MIN   PTS   FGM   FGA  3PM  3PA  FTM  \
0     Allen Iverson  PHI  2000  71  42.0  31.1  10.7  25.5  1.4  4.3  8.2   
1  Jerry Stackhouse  DET  2000  80  40.2  29.8   9.7  24.1  2.1  5.9  8.3   
2  Shaquille O'Neal  LAL  2000  74  39.5  28.7  11.0  19.2  0.0  0.0  6.7   
3       Kobe Bryant  LAL  2000  68  40.9  28.5  10.3  22.2  0.9  2.9  7.0   
4      Vince Carter  TOR  2000  75  39.7  27.6  10.2  22.1  2.2  5.3  5.1   

    FTA  OREB  DREB  AST  STL  BLK  TOV  
0  10.1   0.7   3.1  4.6  2.5  0.3  3.3  
1  10.1   1.2   2.7  5.1  1.2  0.7  4.1  
2  13.1   3.9   8.8  3.7  0.6  2.8  2.9  
3   8.2   1.5   4.3  5.0  1.7  0.6  3.2  
4   6.7   2.3   3.2  3.9  1.5  1.1  2.2

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
    <font face="vazir" color="#0099cc">
        قسمت اول
    </font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
         لیستی از بازیکن‌ها و فصل‌ها را بدست آورید که در آن فصل، آن بازیکن در حداقل ۲ تا از ۵ ملاک اصلی آمار دو رقمی (بزرگتر مساوی ۱۰) ثبت کرده باشد. 
        <br>
        منظور از ۵ ملاک اصلی میانگین امتیاز، میانگین تعداد ریباند (مجموع ریباند در دفاع و حمله)، میانگین تعداد پاس‌های منجر به گل، میانگین تعداد بلاک‌ها و میانگین تعداد توپ‌ربایی‌ها است.
    </font>
    <br>
    <font face="vazir" color="red">
        خروجی:
    </font>
    نتیجه‌ی درخواست را در دیتافریمی با نام <code>double_double</code> قرار دهید.
    این دیتافریم باید شامل ۷ ستون باشد که ستون‌ها از چپ به راست به ترتیب نام بازیکن، سال، میانگین امتیاز، میانگین تعداد پاس‌های منجر به گل، میانگین تعداد ریباند‌ها، میانگین تعداد بلاک‌ها و میانگین تعداد توپ‌ربایی‌ها باشند. همچنین کل دیتافریم باید به ترتیب برحسب سال و نام بازیکن به صورت صعودی مرتب باشد. دو سطر اول این دیتافریم به شکل زیر است:
</p>

<center>
<div style="line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|PLAYER|YEAR|PTS|AST|REB|STL|BLK|
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|Antonio Davis|2000|13.7|1.4|10.1|0.3|1.9|
|Antonio McDyess|2000|20.8|2.1|12.0|0.6|1.5|
    
</font>
</div>
</center>

In [30]:
df['REB'] = df['OREB'] + df['DREB']
double_double = df[['PLAYER', 'YEAR', 'PTS', 'AST', 'REB', 'STL', 'BLK']]
double_double

PLAYER  YEAR   PTS  AST   REB  STL  BLK
0        Allen Iverson  2000  31.1  4.6   3.8  2.5  0.3
1     Jerry Stackhouse  2000  29.8  5.1   3.9  1.2  0.7
2     Shaquille O'Neal  2000  28.7  3.7  12.7  0.6  2.8
3          Kobe Bryant  2000  28.5  5.0   5.8  1.7  0.6
4         Vince Carter  2000  27.6  3.9   5.5  1.5  1.1
...                ...   ...   ...  ...   ...  ...  ...
1825     Marcin Gortat  2009   3.6  0.2   4.2  0.2  0.9
1826       Jeff Teague  2009   3.2  1.7   1.0  0.5  0.2
1827       Kurt Thomas  2009   3.0  0.7   4.2  0.4  0.7
1828   DeMarre Carroll  2009   2.9  0.5   2.1  0.4  0.1
1829      Joel Anthony  2009   2.7  0.2   3.1  0.3  1.4

[1830 rows x 7 columns]

In [31]:
for column in double_double.columns[2:]:
    double_double[f'{column}_c'] = double_double[column] >= 10

C:\Users\masou\AppData\Local\Temp\ipykernel_10696\3364449345.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  double_double[f'{column}_c'] = double_double[column] >= 10
C:\Users\masou\AppData\Local\Temp\ipykernel_10696\3364449345.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  double_double[f'{column}_c'] = double_double[column] >= 10
C:\Users\masou\AppData\Local\Temp\ipykernel_10696\3364449345.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using

In [32]:
double_double = double_double[double_double[['PTS_c', 'AST_c', 'REB_c', 'STL_c', 'BLK_c']].sum(axis=1) > 1][['PLAYER', 'YEAR', 'PTS', 'AST', 'REB', 'STL', 'BLK']]
double_double = double_double.sort_values(by=['YEAR', 'PLAYER']).reset_index(drop=True)
double_double

PLAYER  YEAR   PTS   AST   REB  STL  BLK
0     Antonio Davis  2000  13.7   1.4  10.1  0.3  1.9
1   Antonio McDyess  2000  20.8   2.1  12.0  0.6  1.5
2      Chris Webber  2000  27.1   4.2  11.1  1.3  1.7
3   Dikembe Mutombo  2000  10.0   1.0  13.5  0.4  2.7
4       Elton Brand  2000  20.1   3.2  10.1  1.0  1.6
..              ...   ...   ...   ...   ...  ...  ...
83   Gerald Wallace  2009  18.2   2.1  10.1  1.5  1.1
84       Steve Nash  2009  16.5  11.0   3.3  0.5  0.1
85       Tim Duncan  2009  17.9   3.2  10.1  0.6  1.5
86      Troy Murphy  2009  14.6   2.1  10.2  1.0  0.5
87    Zach Randolph  2009  20.8   1.8  11.8  1.0  0.4

[88 rows x 7 columns]

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
    <font face="vazir" color="#0099cc">
        قسمت دوم
    </font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
         باارزش‌ترین بازیکنان این ۱۰ فصل را مشخص کنید.
        ارزش یک بازیکن در یک فصل با استفاده از فرمول زیر بدست می‌آید:
    </font>
    $$ (PTS + OREB + DREB + AST + STL + BLK) - (TOV + Missed FG + Missed FT) $$
    منظور از <code>Missed FG</code> و <code>Missed FT</code> به ترتیب میانگین تعداد پرتاب‌های ۲ امتیازی و آزاد از دست رفته است.  
    <br>
    ارزش یک بازیکن در کل ۱۰ فصل برابر است با میانگین ارزش او در هر فصل.
    <br>
    <font face="vazir" color="red">
        خروجی:
    </font>
    نتیجه‌ی درخواست را در دیتافریمی با نام <code>best_player</code> قرار دهید.
    این دیتافریم شامل دو ستون است که یکی نام بازیکن و دومی ارزش آن بازیکن است که تا دو رقم اعشار باید گرد شود. همچنین این دیتافریم باید بر اساس ارزش به صورت نزولی و سپس بر اساس نام بازیکن به‌صورت صعودی مرتب باشد.
    دو سطر اول این دیتافریم به شکل زیر است:
</p>

<center>
<div style="line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|PLAYER|VALUE|
|:------:|:---:|
|Kevin Garnett|29.74|
|LeBron James|27.9|
    
</font>
</div>
</center>

$$ (PTS + OREB + DREB + AST + STL + BLK) - (TOV + Missed FG + Missed FT) $$

In [143]:
df2 = df.copy()

In [144]:
df2.head()

PLAYER TEAM  YEAR  GP   MIN   PTS   FGM   FGA  3PM  3PA  FTM  \
0     Allen Iverson  PHI  2000  71  42.0  31.1  10.7  25.5  1.4  4.3  8.2   
1  Jerry Stackhouse  DET  2000  80  40.2  29.8   9.7  24.1  2.1  5.9  8.3   
2  Shaquille O'Neal  LAL  2000  74  39.5  28.7  11.0  19.2  0.0  0.0  6.7   
3       Kobe Bryant  LAL  2000  68  40.9  28.5  10.3  22.2  0.9  2.9  7.0   
4      Vince Carter  TOR  2000  75  39.7  27.6  10.2  22.1  2.2  5.3  5.1   

    FTA  OREB  DREB  AST  STL  BLK  TOV   REB  
0  10.1   0.7   3.1  4.6  2.5  0.3  3.3   3.8  
1  10.1   1.2   2.7  5.1  1.2  0.7  4.1   3.9  
2  13.1   3.9   8.8  3.7  0.6  2.8  2.9  12.7  
3   8.2   1.5   4.3  5.0  1.7  0.6  3.2   5.8  
4   6.7   2.3   3.2  3.9  1.5  1.1  2.2   5.5

In [145]:
df2['M_FG'] = df2['FGA'] - df2['FGM']
df2['M_FT'] = df2['FTA'] - df2['FTM']

In [146]:
df2.head()

PLAYER TEAM  YEAR  GP   MIN   PTS   FGM   FGA  3PM  3PA  ...  \
0     Allen Iverson  PHI  2000  71  42.0  31.1  10.7  25.5  1.4  4.3  ...   
1  Jerry Stackhouse  DET  2000  80  40.2  29.8   9.7  24.1  2.1  5.9  ...   
2  Shaquille O'Neal  LAL  2000  74  39.5  28.7  11.0  19.2  0.0  0.0  ...   
3       Kobe Bryant  LAL  2000  68  40.9  28.5  10.3  22.2  0.9  2.9  ...   
4      Vince Carter  TOR  2000  75  39.7  27.6  10.2  22.1  2.2  5.3  ...   

    FTA  OREB  DREB  AST  STL  BLK  TOV   REB  M_FG  M_FT  
0  10.1   0.7   3.1  4.6  2.5  0.3  3.3   3.8  14.8   1.9  
1  10.1   1.2   2.7  5.1  1.2  0.7  4.1   3.9  14.4   1.8  
2  13.1   3.9   8.8  3.7  0.6  2.8  2.9  12.7   8.2   6.4  
3   8.2   1.5   4.3  5.0  1.7  0.6  3.2   5.8  11.9   1.2  
4   6.7   2.3   3.2  3.9  1.5  1.1  2.2   5.5  11.9   1.6  

[5 rows x 21 columns]

In [147]:
df2['VALUE'] = (df2['PTS']+df2['OREB']+df2['DREB']+df2['AST']+df2['STL']+df2['BLK']) - (df2['TOV']+df2['M_FG']+df2['M_FT'])
df2.head()

PLAYER TEAM  YEAR  GP   MIN   PTS   FGM   FGA  3PM  3PA  ...  \
0     Allen Iverson  PHI  2000  71  42.0  31.1  10.7  25.5  1.4  4.3  ...   
1  Jerry Stackhouse  DET  2000  80  40.2  29.8   9.7  24.1  2.1  5.9  ...   
2  Shaquille O'Neal  LAL  2000  74  39.5  28.7  11.0  19.2  0.0  0.0  ...   
3       Kobe Bryant  LAL  2000  68  40.9  28.5  10.3  22.2  0.9  2.9  ...   
4      Vince Carter  TOR  2000  75  39.7  27.6  10.2  22.1  2.2  5.3  ...   

   OREB  DREB  AST  STL  BLK  TOV   REB  M_FG  M_FT  VALUE  
0   0.7   3.1  4.6  2.5  0.3  3.3   3.8  14.8   1.9   22.3  
1   1.2   2.7  5.1  1.2  0.7  4.1   3.9  14.4   1.8   20.4  
2   3.9   8.8  3.7  0.6  2.8  2.9  12.7   8.2   6.4   31.0  
3   1.5   4.3  5.0  1.7  0.6  3.2   5.8  11.9   1.2   25.3  
4   2.3   3.2  3.9  1.5  1.1  2.2   5.5  11.9   1.6   23.9  

[5 rows x 22 columns]

In [148]:
best_player= df2.groupby(['PLAYER']).agg({
    'VALUE': 'mean'
})


In [149]:
best_player['VALUE'] = best_player['VALUE'].round(2)
best_player = best_player.sort_values(by=['VALUE', 'PLAYER'], ascending=[False, True]).reset_index()
best_player

PLAYER  VALUE
0           Kevin Garnett  29.74
1            LeBron James  27.90
2              Chris Paul  26.57
3        Shaquille O'Neal  26.55
4             Dwyane Wade  26.12
..                    ...    ...
566          Quincy Douby   3.60
567            Steve Kerr   3.60
568      Brian Scalabrine   3.30
569  Nikoloz Tskitishvili   3.30
570          Sam Mitchell   3.15

[571 rows x 2 columns]

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
    <font face="vazir" color="#0099cc">
        قسمت سوم
    </font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
         در هر سال چه تیمی بیشترین میانگین امتیاز را ثبت کرده است؟
        <br>
        توجه داشته باشید که میانگین امتیاز یک تیم برابر است با مجموع میانگین امتیاز بازیکن‌هایش.
    </font>
    <br>
    <font face="vazir" color="red">
        خروجی:
    </font>
    نتیجه‌ی درخواست را در دیتافریمی با نام <code>max_PTS_of_year</code> قرار دهید.
    این دیتافریم باید شامل سه ستون باشد که ستون‌ها از چپ به راست به ترتیب سال، نام تیم و میانگین امتیاز است که امتیاز‌ها باید تا دو رقم اعشار گرد شوند. همچنین کل دیتافریم باید برحسب سال به صورت صعودی مرتب باشد. دو سطر اول این دیتافریم به شکل زیر است:
</p>

<center>
<div style="line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|YEAR|TEAM|PTS|
|:------:|:---:|:---:|
|2000|SAC|100.2|
|2001|LAL|95.9|
    
</font>
</div>
</center>

In [123]:
df3 = df.copy()
df3.head()

PLAYER TEAM  YEAR  GP   MIN   PTS   FGM   FGA  3PM  3PA  FTM  \
0     Allen Iverson  PHI  2000  71  42.0  31.1  10.7  25.5  1.4  4.3  8.2   
1  Jerry Stackhouse  DET  2000  80  40.2  29.8   9.7  24.1  2.1  5.9  8.3   
2  Shaquille O'Neal  LAL  2000  74  39.5  28.7  11.0  19.2  0.0  0.0  6.7   
3       Kobe Bryant  LAL  2000  68  40.9  28.5  10.3  22.2  0.9  2.9  7.0   
4      Vince Carter  TOR  2000  75  39.7  27.6  10.2  22.1  2.2  5.3  5.1   

    FTA  OREB  DREB  AST  STL  BLK  TOV   REB  
0  10.1   0.7   3.1  4.6  2.5  0.3  3.3   3.8  
1  10.1   1.2   2.7  5.1  1.2  0.7  4.1   3.9  
2  13.1   3.9   8.8  3.7  0.6  2.8  2.9  12.7  
3   8.2   1.5   4.3  5.0  1.7  0.6  3.2   5.8  
4   6.7   2.3   3.2  3.9  1.5  1.1  2.2   5.5

In [124]:
max_PTS_of_year_pre = df3.groupby(['YEAR', 'TEAM']).agg({
    'PTS': 'sum'
}).reset_index()
max_PTS_of_year_pre

YEAR TEAM   PTS
0    2000  ATL  52.1
1    2000  BOS  72.5
2    2000  CHH  77.3
3    2000  CHI  50.6
4    2000  CLE  57.0
..    ...  ...   ...
291  2009  SAC  99.6
292  2009  SAS  83.4
293  2009  TOR  78.7
294  2009  UTA  72.7
295  2009  WAS  43.5

[296 rows x 3 columns]

In [128]:
max_PTS_of_year = max_PTS_of_year_pre.groupby(['YEAR'], as_index=False).agg({
    'PTS':'max'
})
max_PTS_of_year

YEAR    PTS
0  2000  100.2
1  2001   95.9
2  2002   97.0
3  2003   96.0
4  2004  106.2
5  2005   96.8
6  2006  105.4
7  2007  110.7
8  2008   97.3
9  2009  102.5

In [129]:
max_PTS_of_year = max_PTS_of_year.join(max_PTS_of_year_pre.set_index(['PTS', 'YEAR']), on=['PTS', 'YEAR'])[['YEAR', 'TEAM', 'PTS']]
max_PTS_of_year['PTS'] = max_PTS_of_year['PTS'].round(2)
max_PTS_of_year

YEAR TEAM    PTS
0  2000  SAC  100.2
1  2001  LAL   95.9
2  2002  ORL   97.0
3  2003  DEN   96.0
4  2004  BOS  106.2
5  2005  WAS   96.8
6  2006  PHX  105.4
7  2007  DEN  110.7
8  2008  DEN   97.3
9  2009  OKC  102.5

<h2 dir=rtl align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
<b>سلول جواب‌ساز</b>
</font>
</h2>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazir; font-size:medium">
<font face="vazir" size=3>
    برای ساخته‌شدن فایل <code>result.zip</code> سلول زیر را اجرا کنید. توجه داشته باشید که پیش از اجرای سلول زیر تغییرات اعمال شده در نت‌بوک را ذخیره کرده باشید (<code>ctrl+s</code>) در غیر این صورت، در پایان مسابقه نمره شما به صفر تغییر خواهد کرد.
    <br>
    همچنین اگر از کولب برای اجرای این فایل نوت‌بوک استفاده می‌کنید، قبل از ارسال فایل <code>result.zip</code>، آخرین نسخه‌ی نوت‌بوک خود را دانلود کرده و داخل فایل ارسالی قرار دهید.
</font>

In [103]:
import zipfile
import os

if not os.path.exists(os.path.join(os.getcwd(), 'NBA.ipynb')):
    %notebook -e NBA.ipynb

double_double.to_csv('double_double.csv', index=False)
best_player.to_csv('best_player.csv', index=False)
max_PTS_of_year.to_csv('max_PTS_of_year.csv', index=False)

def compress(file_names):
    print("File Paths:")
    print(file_names)
    compression = zipfile.ZIP_DEFLATED
    with zipfile.ZipFile("result.zip", mode="w") as zf:
        for file_name in file_names:
            zf.write('./' + file_name, file_name, compress_type=compression)

file_names = ['double_double.csv', 'best_player.csv', 'max_PTS_of_year.csv', 'NBA.ipynb']
compress(file_names)

File Paths:
['double_double.csv', 'best_player.csv', 'max_PTS_of_year.csv', 'NBA.ipynb']
